### ISSUE TITLE
Gradient accumulation in the inner loop

### ISSUE THREAD

**Issue by jkang1640** [learnables/learn2learn#394]
Hello, 

I am using learn2learn for seq2seq models and thank you very much for the awesome library! 

Because of the gpu limit, I'd like to use gradint accumulation. For the outer loop, there is no problem implementing this but inside innder loop, I don't know how to do this efficiently since `adapt` step includes both backward and update step. 

 Here's my code blow. This is not efficient way at all since I keep accumulating `outputs.loss` instead of accumulating gradients only. Therefore, computation graphs are not released till I call `learner.adapt(loss)`. 

This causes OOM when I run my code. Is there any way to apply gradient accumulation efficiently for the inner loop? Thank you very much for your help! 

```
    def inner_loop(learner, support_dl, adaptation_steps):

        total_loss = []
        for step in range(adaptation_steps):
            loss = 0.0

            for inputs in support_dl:
                outputs = learner(**inputs)
                curr_loss = outputs.loss
                loss += (curr_loss * support_dl.batch_size / len(support_dl.dataset))

            learner.adapt(loss)
            total_loss.append(loss.item())

        return learner, sum(total_loss) / len(total_loss)
```

**Comment by jkang1640** [comment]
I checked maml.py and modified my code as below. Could anybody tell me if this is right way to do gradient accumulation inside the inner loop? Thank you. 

```
    from learn2learn.algorithms import maml_update
    from torch import autograd 

    def inner_loop(learner, support_dl, fast_lr, adaptation_steps): 
        
        all_loss = []
        total_grads = (0, 0)

        for step in range(adaptation_steps):
            loss = 0.0

            for inputs in support_dl:
                outputs = learner(**inputs)
                curr_loss = outputs.loss
                grads = autograd.grad(curr_loss, learner.parameters(), create_graph=False) # False for FOMAML
                temp_grad1 = torch.add(total_grads[0], grads[0]) # gradient accum
                temp_grad2 = torch.add(total_grads[1], grads[1]) # gradient accum
                total_grads = (temp_grad1, temp_grad2)

                loss += (curr_loss.item() * support_dl.batch_size / len(support_dl.dataset)) # accum loss for logging 
                del outputs, curr_loss, grads

            maml_update(learner, lr=fast_lr, grads=total_grads) # update learner parameters with accumulated gradients 
            all_loss.append(loss)

        return learner, sum(all_loss) / len(all_loss)
```

**Comment by jkang1640** [comment]
After reading the code maml.py, I changed my code this way. 

Could anybody tell me if it seems right..? Just one more thing, I have no idea when to use `learner.module` or `learner`. For example, which one should I return at the end of the code? Which one to pass on to `maml_update`?  
```
    def inner_loop(learner, support_dl, fast_lr, adaptation_steps):
       

        all_loss = []
        # initiliaze gradients to zero for all parameters

        accum_grads = tuple(torch.zeros(param.size(), device=device) for param in learner.module.parameters())

        for step in range(adaptation_steps):
            loss = 0.0

            for inputs in support_dl:
                outputs = learner(**inputs)
                curr_loss = outputs.loss
                curr_grads = autograd.grad(curr_loss, learner.module.parameters(), create_graph=False) # False for FOMAML
                accum_grads = tuple(torch.add(accum_grads[i], curr_grads[i]) for i in range(len(curr_grads)))
                loss += (curr_loss.item() * support_dl.batch_size / len(support_dl.dataset)) # accum loss for logging
                del outputs, curr_loss, curr_grads

            maml_update(learner, lr=fast_lr, grads=accum_grads) # update learner parameters with accumulated gradients
            all_loss.append(loss)

        return learner, sum(all_loss) / len(all_loss)
```

**Comment by seba-1511** [comment]
Hello @jkang1640,

Sequence models are notoriously tricky to train with MAML as it's easy to run into OOM like you did.

Unfortunately there's no easy fix for when you're trying to accumulate gradients in the inner loop. My advice is to either pretrain and freeze some of the parameters of the model as in ANIL (to reduce memory consumption during meta-learning), or maybe go for a first-order algorithm (FOMAM, ANIL).

I'll close this for now.
